src/rust/ide/view/graph-editor/src/component/tooltip.rs

```diff
@@ -0,0 +1,224 @@
+        Self{text}
+    }
+    /// Create a `TooltipUpdate` that unsets the label of the tooltip.
+    pub fn unset_label() -> Self {
+        let text = Some(StyleValue::new(None));
+        Self{text}
+    }
+
+    fn has_text(&self) -> bool {
+        if let Some(style_value) = self.text.as_ref() {
+            if let Some(inner) = style_value.value.as_ref() {
+                return inner.is_some()
+            }
+        }
+        false
+    }
+}
+
+
+
+// ==============
+// === Offset ===
+// ==============
+
+#[derive(Clone,Copy,Debug)]
+#[allow(missing_docs)]
+/// Indicates the position of the tooltip relative to the cursor location.
+pub enum Offset {
+    Top, Bottom, Left, Right
+}
+
+impl Default for Offset {
+    fn default() -> Self {
+        Offset::Bottom
+    }
+}
+
+ensogl::define_endpoints! {
+    Input {
+        set_style    (Style),
+        set_location (Vector2),
+        set_offset   (Offset)
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+#[derive(Clone,Debug)]
+struct Model {
+    tooltip : TextLabel,
+    root    : display::object::Instance,
+}
+
+impl Model {
+    fn new(app:&Application) -> Self {
+        let logger  = Logger::new("TooltipModel");
+        let tooltip = TextLabel::new(app.clone_ref());
+        let root    = display::object::Instance::new(&logger);
+        root.add_child(&tooltip);
+
+        let tgt_layer = &app.display.scene().layers.cursor;
+        tgt_layer.add_exclusive(&tooltip);
+
+        Self{tooltip,root}
+    }
+
+    fn set_location(&self, position:Vector2, size:Vector2, layout: Offset) {
+        let layout_offset = match layout {
+            Offset::Top => {
+                Vector2::new(0.0,size.y * 0.5)
+            }
+            Offset::Bottom => {
+                Vector2::new(0.0,-size.y * 0.5)
+            }
+            Offset::Left => {
+              Vector2::new(-size.x / 2.0,0.0)
+            }
+            Offset::Right => {
+               Vector2::new(size.x / 2.0,0.0)
+            }
+        };
+
+        let base_positions = position.xy();
+        self.tooltip.set_position_xy(base_positions + layout_offset)
+    }
+
+    fn set_style(&self, update:&Style) {
+        if let Some(style) = update.text.as_ref() {
+            let text = style.value.clone().flatten();
+            if let Some(text) = text {
+                self.tooltip.frp.set_content(text)
+            }
+        }
+    }
+
+    fn set_visibility(&self, visible:bool) {
+        if visible {
+            self.root.add_child(&self.tooltip)
+        } else {
+            self.tooltip.unset_parent()
+        }
+    }
+
+    fn set_opacity(&self, value:f32) {
+        self.tooltip.frp.set_opacity(value);
+    }
+}
+
+
+
+// ===============
+// === Tooltip ===
+// ===============
+
+/// Tooltip component that can show extra information about other UI components.
+#[derive(Clone,CloneRef,Debug)]
+pub struct Tooltip {
+    model   : Rc<Model>,
+    /// Public FRP api.
```

Comment:
rm this doc, it doesnt tell anything more than source code. I'd always hide such docs as they make the codeless redable imo (I know this may be just my opinion, but Id be thankful for that)